### PR TITLE
[X86][GlobalISel] Use GIntrinsic during custom intrinsic selection

### PR DESCRIPTION
**Pull request by e-kud**

### CODE REVIEW

**Review by tschuett — APPROVED** [review]

File: llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp [hunk 1]
```diff
@@ -1843,7 +1843,7 @@ bool X86InstructionSelector::selectIntrinsicWSideEffects(
   assert(I.getOpcode() == TargetOpcode::G_INTRINSIC_W_SIDE_EFFECTS &&
          "unexpected instruction");
 
-  if (I.getOperand(0).getIntrinsicID() != Intrinsic::trap)
+  if (cast<GIntrinsic>(I).getIntrinsicID() != Intrinsic::trap)
```
Comment (arsenm) [inline llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp:1846]: generic intrinsics should not use G_INTRINSIC, this shouldn't have been reachable

Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp:1846]: That's interesting. It seems that `llvm.trap` is a special case as well as `llvm.debugtrap` and `llvm.ubsantrap`. Comparing to other intrinsics we can't lower it into combination of GlobalISel opcodes. So it is alive until instruction selection and then we custom select it.

If we take a look at SelectionDAG it lowers the intrinsic into `ISD::TRAP`. It seems that we can create `G_TRAP/G_DEBUGTRAP/G_UBSANTRAP` opcodes and lower intrinsics during `irtranslator`. I see that only 3 targets have a custom lowering for traps: AArch64, X86 and RISCV. So it shouldn't be a problem to migrate them all at once to proper lowering. What do you think, does it worth it?

Comment (arsenm) [reply llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp:1846]: Yes, that's how I would expect this to work

**Review by arsenm — APPROVED** [review]
This is fine until the new G_* instructions are added

**Comment by e-kud** [comment]
@tschuett @arsenm thank you for the review! The PR with new intrinsic opcodes has been merged instead: https://github.com/llvm/llvm-project/pull/84941

**Pull request closed without merging**